### ISSUE TITLE
2.0.X on OS X El Capitan

### ISSUE THREAD

**Issue by darach** [warmcat/libwebsockets#554]
Working beautifully on debian x86-64 and OpenWRT mips32r2, but
SSL ( OpenSSL ) connections on OS X El Capitan are hanging to
the same address:

For example:

```
$ libwebsockets-test-ping some.website -p 443 --ssl
```

```
89 693.994240   X.X.X.X -> Y.Y.Y.Y TCP 78 53478 → 443 [SYN] Seq=0 Win=65535 Len=0 MSS=1460 WS=32 TSval=1320909215 TSecr=0 SACK_PERM=1
990 694.004010 Y.Y.Y.Y -> X.X.X.X   TCP 74 443 → 53478 [SYN, ACK] Seq=0 Ack=1 Win=28960 Len=0 MSS=1460 SACK_PERM=1 TSval=2781464451 TSecr=1320909215 WS=128
991 694.004095   X.X.X.X -> Y.Y.Y.Y TCP 66 53478 → 443 [ACK] Seq=1 Ack=1 Win=131744 Len=0 TSval=1320909225 TSecr=2781464451
992 694.005735   X.X.X.X -> Y.Y.Y.Y SSL 192 Client Hello
993 694.017135 Y.Y.Y.Y -> X.X.X.X  TCP 66 443 → 53478 [ACK] Seq=1 Ack=127 Win=29056 Len=0 TSval=2781464454 TSecr=1320909226
994 694.017304 Y.Y.Y.Y -> X.X.X.X   TCP 66 443 → 53478 [RST, ACK] Seq=1 Ack=127 Win=29056 Len=0 TSval=2781464454 TSecr=1320909226

// hangs ...
```

Debug output:

```
[2016/06/03 12:29:02:1149] NOTICE: Initial logging level 7
[2016/06/03 12:29:02:1149] NOTICE: Libwebsockets version: 2.0.1 myself@self.local-v2.0.2
[2016/06/03 12:29:02:1149] NOTICE: IPV6 not compiled in
[2016/06/03 12:29:02:1149] NOTICE: libev support not compiled in
[2016/06/03 12:29:02:1149] NOTICE: libuv support not compiled in
[2016/06/03 12:29:02:1150] NOTICE:  Threads: 1 each 8000 fds
[2016/06/03 12:29:02:1151] NOTICE:  mem: platform fd map: 64000 bytes
[2016/06/03 12:29:02:1151] NOTICE:  Compiled with OpenSSL support
[2016/06/03 12:29:02:1158] NOTICE: Creating Vhost 'default' port -1, 1 protocols
[2016/06/03 12:29:02:1162] NOTICE:  mem: per-conn:          488 bytes + protocol rx buf
[2016/06/03 12:29:02:1162] NOTICE:  canonical_hostname = self.local
[2016/06/03 12:29:02:1162] NOTICE: Connecting to some.website:443...
[2016/06/03 12:29:02:1200] WARN: getpeername: Socket is not connected
[2016/06/03 12:29:02:1200] NOTICE: libwebsockets test server ping - license LGPL2.1+SLE
[2016/06/03 12:29:02:1200] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
Websocket PING  () 64 bytes of data.
[2016/06/03 12:29:02:1227] NOTICE: lws_protocol_init
^Z
```

Are there known issues on recent OS X that don't impact Linux environments? If there are known issues are they documented in some way? I've tried multiple versions from 1.6.2 ( via homebrew ) and local builds of 1.7.x up to 2.0.2 and keep hitting the same issue on OS X, but no issues on Linux or embedded Linux environments.

**Comment by lws-team** [comment]
There are no "known problems" since nobody else reported trouble and I am disinclined to give Apple money so I can discover their problems myself.  If we do get to know about problems we try to fix them if it's on our side.

My guess would be this is related to the SSL library on the OS, I read before that Apple's OpenSSL was very old.  Otherwise Debian and generic BSD are largely going to act the same I would have thought.  It's at least worth looking for and reporting the OS SSL version.

Also on the default test apps, the ciphers + key exchanges are restricted to the really good ones, with PFS.  If "El Capitan" implies it's not recent, maybe it does not have the EC support needed for that (in which case try hacking the info.ssl_cipher_list to be "ALL").

**Comment by darach** [comment]
OpenSSL version being used is 1.0.2h. Tried setting the cipher list to 'ALL' as suggested but
that didn't help. Also, I should note the server I'm hitting has a restricted set so I've also enabled
ECDH. Also, everything seems fine without SSL, so this is indeed some kind of SSL issue.

I'll see if I can get some more debug output. Interestingly 1.7.8 reports a connection failure but 2.0.2
hangs. I'll see if I can dig up some more detail...

**Comment by lws-team** [comment]
If you're using libuv / lwsws then a 'hang' can indicate a fatal signal... if you can cleanly ^C the process, then no problem.  Otherwise attach `sudo gdb -p XXX` and use `bt` to see where it hanged.

You might want to make sure you build in DEBUG mode

cmake .. -DCMAKE_BUILD_TYPE=DEBUG

**Comment by darach** [comment]
Tried with a few versions of LWS and openssl, similar outcomes. Also doesn't seem to matter
whether --ssl is specified or not on the commandline:

```
(lldb) run some.website
Process 4739 launched: ‘./libwebsockets-test-ping' (x86_64)
[2016/06/03 14:18:28:9733] NOTICE: Initial logging level 7
[2016/06/03 14:18:28:9733] NOTICE: Libwebsockets version: 2.0.0 self@self.local-v2.0.0-62-g4889566
[2016/06/03 14:18:28:9733] NOTICE: IPV6 not compiled in
[2016/06/03 14:18:28:9733] NOTICE: libev support not compiled in
[2016/06/03 14:18:28:9733] NOTICE: libuv support not compiled in
[2016/06/03 14:18:28:9734] NOTICE:  Threads: 1 each 8000 fds
[2016/06/03 14:18:28:9735] NOTICE:  mem: platform fd map: 64000 bytes
[2016/06/03 14:18:28:9735] NOTICE:  Compiled with OpenSSL support
[2016/06/03 14:18:28:9735] NOTICE:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2016/06/03 14:18:28:9735] NOTICE: Creating Vhost 'default' port -1, 1 protocols
[2016/06/03 14:18:28:9735] NOTICE:  mem: per-conn:          496 bytes + protocol rx buf
[2016/06/03 14:18:28:9735] NOTICE:  canonical_hostname = self.local
[2016/06/03 14:18:28:9735] NOTICE: Connecting to some.website:7681...
[2016/06/03 14:18:28:9995] WARN: getpeername: Socket is not connected
[2016/06/03 14:18:28:9996] NOTICE: libwebsockets test server ping - license LGPL2.1+SLE
[2016/06/03 14:18:28:9996] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
Websocket PING  () 64 bytes of data.
[2016/06/03 14:18:29:0008] NOTICE: lws_protocol_init

Process 4739 stopped
* thread #1: tid = 0x1fca461, 0x00007fff924452a2 libsystem_kernel.dylib`poll + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff924452a2 libsystem_kernel.dylib`poll + 10
libsystem_kernel.dylib`poll:
->  0x7fff924452a2 <+10>: jae    0x7fff924452ac            ; <+20>
    0x7fff924452a4 <+12>: movq   %rax, %rdi
    0x7fff924452a7 <+15>: jmp    0x7fff9243e7f2            ; cerror
    0x7fff924452ac <+20>: retq
(lldb) bt
* thread #1: tid = 0x1fca461, 0x00007fff924452a2 libsystem_kernel.dylib`poll + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff924452a2 libsystem_kernel.dylib`poll + 10
    frame #1: 0x000000010001ac00 libwebsockets-test-ping`lws_plat_service_tsi(context=0x0000000101000000, timeout_ms=1, tsi=0) + 416 at lws-plat-unix.c:148
    frame #2: 0x000000010001aedd libwebsockets-test-ping`lws_plat_service(context=0x0000000101000000, timeout_ms=1) + 29 at lws-plat-unix.c:199
    frame #3: 0x0000000100008a9b libwebsockets-test-ping`lws_service(context=0x0000000101000000, timeout_ms=1) + 27 at service.c:1124
    frame #4: 0x0000000100002485 libwebsockets-test-ping`main(argc=2, argv=0x00007fff5fbff0d8) + 2165 at test-ping.c:531
    frame #5: 0x00007fff8ede05ad libdyld.dylib`start + 1
    frame #6: 0x00007fff8ede05ad libdyld.dylib`start + 1
(lldb) frame select 1
frame #1: 0x000000010001ac00 libwebsockets-test-ping`lws_plat_service_tsi(context=0x0000000101000000, timeout_ms=1, tsi=0) + 416 at lws-plat-unix.c:148
   145
   146      timeout_ms = lws_service_adjust_timeout(context, timeout_ms, tsi);
   147
-> 148      n = poll(pt->fds, pt->fds_count, timeout_ms);
   149
   150  #ifdef LWS_OPENSSL_SUPPORT
   151      if (!pt->rx_draining_ext_list &&
(lldb) frame select 2
frame #2: 0x000000010001aedd libwebsockets-test-ping`lws_plat_service(context=0x0000000101000000, timeout_ms=1) + 29 at lws-plat-unix.c:199
   196  LWS_VISIBLE int
   197  lws_plat_service(struct lws_context *context, int timeout_ms)
   198  {
-> 199      return lws_plat_service_tsi(context, timeout_ms, 0);
   200  }
   201
   202  LWS_VISIBLE int
(lldb) frame select 3
frame #3: 0x0000000100008a9b libwebsockets-test-ping`lws_service(context=0x0000000101000000, timeout_ms=1) + 27 at service.c:1124
   1121 LWS_VISIBLE int
   1122 lws_service(struct lws_context *context, int timeout_ms)
   1123 {
-> 1124     return lws_plat_service(context, timeout_ms);
   1125 }
   1126
   1127 LWS_VISIBLE int
(lldb) frame select 4
frame #4: 0x0000000100002485 libwebsockets-test-ping`main(argc=2, argv=0x00007fff5fbff0d8) + 2165 at test-ping.c:531
   528          if (!interval_us)
   529              n = lws_service(context, 0);
   530          else
-> 531              n = lws_service(context, 1);
   532      }
   533
   534      /* stats */
(lldb)
```

Getting link failures with a fresh build of openssl. If you have a cmake invocation handy to
direct cmake to use the right openssl library that might help isolate things with a debug build
of openssl. Seems to find the right include dir and openssl bin for generating certs for the test
servers/clients, but not the right lib path... No doubt something I haven't grokked right...

**Comment by lws-team** [comment]
About cmake and non-distro ssl, have a look at note 4 here

https://github.com/warmcat/libwebsockets/blob/master/README.build.md#building-on-unix

Reviewing what you wrote thismorning and looking at the trace, there is no 'hang' in the conventional sense.  It's received a RST so the socket is closed it seems.  Then he just sits in poll() with nothing happening.

Maybe make sure you build in DEBUG mode and crank up the log level with -d65535

**Comment by darach** [comment]
Thanks for the link, I'll have a look and see what it throws up.

**Comment by namowen** [comment]
Here is a build script I use on Ubuntu to create test servers and clients linked statically with the openssl libraries. Additionally I've had to adapt CMakeLists.txt on line 941:
build script:
```
#!/bin/bash
cd /home/bzuber/dev/ubuntu/libwebsockets-2.0.1
rm -rf build
mkdir build
cd build
cmake \
		-DCMAKE_BUILD_TYPE=Debug \
		-DCMAKE_EXE_LINKER_FLAGS="-ldl" \
		-DLWS_SSL_CLIENT_USE_OS_CA_CERTS:BOOL=OFF \
		-DLWS_WITH_SSL=ON \
		-DCMAKE_INCLUDE_DIRECTORIES_PROJECT_BEFORE=/usr/local/ssl \
		-DOPENSSL_VERSION=1.0.2h \
		-DOPENSSL_ROOT_DIR=/usr/local/ssl \
		-DOPENSSL_LIBRARIES=/user/local/ssl/lib \
		-DOPENSSL_INCLUDE_DIR=/usr/local/ssl/include \
		-DOPENSSL_EXECUTABLE=/usr/local/ssl/bin/openssl \
		-DOPENSSL_CRYPTO_LIBRARY=/usr/local/ssl/lib/libcrypto.a \
		-DOPENSSL_SSL_LIBRARY=/usr/local/ssl/lib/libssl.a \
		/home/bzuber/dev/ubuntu/libwebsockets-2.0.1
make
sudo make install

```
CMakeLists.txt, line 941:
```
			target_link_libraries(${TEST_NAME} websockets /usr/local/ssl/lib/libcrypto.a /usr/local/ssl/lib/libssl.a dl)
#			target_link_libraries(${TEST_NAME} websockets)

```
Maybe this helps. I'm not a cmake specialist, but I think the changes in CMakeLists.txt could be somehow be made dependable to the flag OPENSSL_USE_STATIC_LIBS, so it could be set with -DOPENSSL_USE_STATIC_LIBS:BOOL=TRUE. At the moment this flas is ignored/not used by CMakeLists.txt of libwebsockets.

**Comment by JoakimSoderberg** [comment]
You should not hard code the paths liks this... especially since CMake is a multi-platform tool, those paths won't work on Windows.. or OSX for that matter.

Also master line 941 is something completely different:
https://github.com/warmcat/libwebsockets/blob/master/CMakeLists.txt#L941

Unfortunately, CMake as I understand does not have a simple way to say to the find scripts "Only return static libraries"... It is kind of up to each and every find script for the libs to take care of that, and none do. But I might just be mistaken :)

In short. Your solution for this is not a viable generic solution to compile libwebsockets with a static openssl

**Comment by darach** [comment]
Some further progress on this resolves the 'hang'. However, now there's a difference between Linux and OS X behaviour:

ERR: server's cert didn't look good, X509_V_ERR = 20

This is occurring on Mac OS X but *not* on Linux where the connection is fine.

The cert isn't the problem. This has been flagged as an lws issue before and recently pinned to a double free on 2.0.x ( https://github.com/warmcat/libwebsockets/issues/541 ) - difference being this is 1.7.8 here and as such the root cause may not be the same - but the error as reported is totally misleading.

**Comment by lws-team** [comment]
> Some further progress on this resolves the 'hang'

This is a pretty abstract piece of news, it gives me no info on what 'progressed', where or why.  This means there was some problem on your side you cleared and we don't need to go into it, or what?

> totally misleading

I don't think so... the code looks at the server cert verify result and tells you if it doesn't look good.  And it reports the ssl library's info on what doesn't look good.

```
n = SSL_get_verify_result(wsi->ssl);
	lws_latency(context, wsi,
		"SSL_get_verify_result LWS_CONNMODE..HANDSHAKE",
						      n, n > 0);

	if (n != X509_V_OK) {
		if ((n == X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT ||
		     n == X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN) && wsi->use_ssl == 2) {
			lwsl_notice("accepting self-signed certificate\n");
		} else {
			lwsl_err("server's cert didn't look good, X509_V_ERR = %d: %s\n",
				 n, ERR_error_string(n, sb));
			lws_ssl_elaborate_error();
			return -1;
		}
	}
```

What do you think that error message should better say so you don't feel 'misled'?

Bear in mind the cert is assessed by the ssl library not in isolation but against CA certs on the system.

If you actually look at what the message tells you, you'll find 20 is X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY

https://github.com/openssl/openssl/blob/master/include/openssl/x509_vfy.h#L105

It sounds less like an inherent osx problem and more like a 'osx doesn't have your CA and can't validate your cert' problem.  It 'works on linux' because that box has the necessary CA cert installed in its OS bundle, if that's what goes on.

**Comment by darach** [comment]
Clearly, "X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY" is far more descriptive
than "20". Simply enhancing the error with that descriptive text rather than "20" would be a great
improvement for most folk who aren't committers to lws.

As to the hang. I stopped developing on OS X. I continued on Linux. I've just come back to OS X
now. So what resolved the hang? I can't pin it. Moving from 1.6.3 to 1.7.8 certainly helped.

Last, but not least... In this case the full cert chain is sent by the server. Does lws add intermediate certs to the locally trusted chain if their parent/root certs are trusted? If so, these are cases where the above error should *not* be issued. If the root is trusted, and the intermediate to leaf certs are supplied, then lws has everything it needs to verify ( or not ) the full chain.

Either way, please close the issue as the 'hang', for which it was opened, is no longer an issue.

**Comment by lws-team** [comment]
If you use the current stable stuff, it will spew more human-readable info since I added code to drain openssl's error buffer before v2.0.  But even before then, the error was telling you the actual problem correctly.

For the rest lws just passes stuff to openssl.  https://libwebsockets.org itself runs on lws using an ehcd cert with intermediates inside, so this works okay in lws on server side.  If you have to give openssl additional certs, you can do it from LWS_CALLBACK_OPENSSL_LOAD_EXTRA_CLIENT_VERIFY_CERTS.

If you find something we ought to fix on lws side, please let us know, but it looks the usual struggle with openssl.
